p2p/router.go

```diff
@@ -0,0 +1,557 @@
+// queues to close, and we wait for this as a signal that goroutines have ended.
+//
+// All message scheduling should be limited to the queue implementations used
+// for channel queues and peer queues. All message sending throughout the router
+// is blocking, and if any messages should be dropped or buffered this is the
+// sole responsibility of the queue, such that we can limit this logic to a
+// single place. There is currently only a FIFO queue implementation that always
+// blocks and never drops messages, but this must be improved with other
+// implementations. The only exception is that all message sending must also
+// select on appropriate channel/queue/router closure signals, to avoid blocking
+// forever on a channel that has no consumer.
+type Router struct {
+	*service.BaseService
+	logger     log.Logger
+	transports map[Protocol]Transport
+	store      *peerStore
+
+	// FIXME: Consider using sync.Map.
+	peerMtx    sync.RWMutex
+	peerQueues map[NodeID]queue
+
+	// FIXME: We don't strictly need to use a mutex for this if we seal the
+	// channels on router start. This depends on whether we want to allow
+	// dynamic channels in the future.
+	channelMtx      sync.RWMutex
+	channelQueues   map[ChannelID]queue
+	channelMessages map[ChannelID]proto.Message
+
+	peerUpdatesCh   chan PeerUpdate
+	peerUpdatesMtx  sync.RWMutex
+	peerUpdatesSubs map[*PeerUpdatesCh]*PeerUpdatesCh // keyed by struct identity (address)
+
+	// stopCh is used to signal router shutdown, by closing the channel.
+	stopCh chan struct{}
+}
+
+// NewRouter creates a new Router, dialing the given peers.
+//
+// FIXME: providing protocol/transport maps is cumbersome in tests, we should
+// consider adding Protocols() to the Transport interface instead and register
+// protocol/transport mappings automatically on a first-come basis.
+func NewRouter(logger log.Logger, transports map[Protocol]Transport, peers []PeerAddress) *Router {
+	router := &Router{
+		logger:          logger,
+		transports:      transports,
+		store:           newPeerStore(),
+		stopCh:          make(chan struct{}),
+		channelQueues:   map[ChannelID]queue{},
+		channelMessages: map[ChannelID]proto.Message{},
+		peerQueues:      map[NodeID]queue{},
+		peerUpdatesCh:   make(chan PeerUpdate, 1),
+		peerUpdatesSubs: map[*PeerUpdatesCh]*PeerUpdatesCh{},
+	}
+	router.BaseService = service.NewBaseService(logger, "router", router)
+
+	for _, address := range peers {
+		if err := router.store.Add(address); err != nil {
+			logger.Error("failed to add peer", "address", address, "err", err)
+		}
+	}
+
+	return router
+}
+
+// OpenChannel opens a new channel for the given message type. The caller must
+// close the channel when done, and this must happen before the router stops.
+func (r *Router) OpenChannel(id ChannelID, messageType proto.Message) (*Channel, error) {
+	// FIXME: NewChannel should take directional channels so we can pass
```

Comment:
I think we can make this change in this PR if you wish.

Reply:
Meh, I'll submit a separate PR, it's a separate change that'll affect some other parts of the code.